crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -363,30 +361,51 @@ impl<'a, Provider: ResolverProvider> Resolver<'a, Provider> {
-                                    IncompatibleWheel::RequiresPython => "no wheels are available that meet your required Python version".to_string(),
-                                    IncompatibleWheel::Tag(tag) => {
-                                        match tag {
-                                            IncompatibleTag::Invalid => "no wheels are available with valid tags".to_string(),
-                                            IncompatibleTag::Python => "no wheels are available with a matching Python implementation".to_string(),
-                                            IncompatibleTag::Abi => "no wheels are available with a matching Python ABI".to_string(),
-                                            IncompatibleTag::Platform => "no wheels are available with a matching platform".to_string(),
+                        UnavailableVersion::IncompatibleDist(incompatibility) => {
+                            match incompatibility {
+                                IncompatibleDist::Wheel(incompatibility) => {
+                                    match incompatibility {
+                                        IncompatibleWheel::NoBinary => "no source distribution is available and using wheels is disabled".to_string(),
+                                        IncompatibleWheel::Tag(tag) => {
+                                            match tag {
+                                                IncompatibleTag::Invalid => "no wheels are available with valid tags".to_string(),
+                                                IncompatibleTag::Python => "no wheels are available with a matching Python implementation".to_string(),
+                                                IncompatibleTag::Abi => "no wheels are available with a matching Python ABI".to_string(),
+                                                IncompatibleTag::Platform => "no wheels are available with a matching platform".to_string(),
+                                            }
+                                        }
+                                        IncompatibleWheel::Yanked(yanked) => match yanked {
+                                            Yanked::Bool(_) => "it was yanked".to_string(),
+                                            Yanked::Reason(reason) => format!(
+                                                "it was yanked (reason: {})",
+                                                reason.trim().trim_end_matches('.')
+                                            ),
+                                        },
+                                        IncompatibleWheel::ExcludeNewer(ts) => match ts {
+                                            Some(_) => "it was published after the exclude newer time".to_string(),
+                                            None => "it has no publish time".to_string()
                                         }
+                                        IncompatibleWheel::RequiresPython(_) => unreachable!(),
```

Comment:
Can `RequiresPython` be refactored to avoid this?

Reply:
You could consider putting this whole block in here instead of using the `continue` and the `unreachable!()`:

```rust
let python_version = requires_python
    .iter()
    .map(PubGrubSpecifier::try_from)
    .fold_ok(Range::full(), |range, specifier| {
        range.intersection(&specifier.into())
    })?;

let package = &next;
for kind in [PubGrubPython::Installed, PubGrubPython::Target] {
    state.add_incompatibility(Incompatibility::from_dependency(
        package.clone(),
        Range::singleton(version.clone()),
        (PubGrubPackage::Python(kind), python_version.clone()),
    ));
}
state.partial_solution.add_decision(next.clone(), version);
continue;
```

Reply:
Then we need to write that whole thing twice which I find less ideal.

My concern is less about the `unreachable!` and more that it suggests a flaw in the abstraction.

Reply:
👍 yeah that makes sense. This would be partially alleviated by a restructuring to avoid duplication as described in https://github.com/astral-sh/uv/pull/1298#discussion_r1516699939